File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Crc8Atm.java
package com.freight.hash;

/** crc8_atm over raw bytes. */
public final class Crc8Atm implements HashAlgorithm {

    @Override
    public String name() {
        return "crc8_atm";
    }

    @Override
    public long apply(byte[] data) {
        long crc = 0L;
        for (int i = 0; i < data.length; i++) {
            crc ^= (data[i] & 0xFF);
            for (int bit = 0; bit < 8; bit++) {
                crc = ((crc & 0x80L) != 0L) ? (((crc << 1) ^ 0x07L) & 0xFFL) : ((crc << 1) & 0xFFL);
            }
        }
        return crc & 0xFFL;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/hashx/crc8_atm.go
package hashx

// Crc8Atm computes the crc8_atm checksum over raw bytes.
func Crc8Atm(data []byte) uint64 {
	crc := uint32(0)
	for _, b := range data {
		crc ^= uint32(b)
		for bit := 0; bit < 8; bit++ {
			if crc&0x80 != 0 {
				crc = ((crc << 1) ^ 0x07) & 0xFF
			} else {
				crc = (crc << 1) & 0xFF
			}
		}
	}
	return uint64(crc & 0xFF)
}
